djhtml/modes.py

```diff
@@ -321,61 +376,127 @@ class DjJS(DjTXT):
     """
 
     RAW_TOKENS = DjTXT.RAW_TOKENS + [
-        r"</script>",
-        r'".*?"',
-        r"'.*?'",
-        r"`.*?`",
-        r"`",
-        r"[\{\[\(\)\]\}]",
+        r"\\.",
         r"//.*",
         r"/\*",
+        r"[\$\w-]+:",
+        r'"(?:\\"|.)*?"',  # "string"
```

Comment:
This regex is wrong, but is just happens to work.

Technically, the `.` is allowed to match the `\` symbol. Thus

```
"world\"
```
is seen as a valid string by this regex, even though the `\` technically escapes the `"`, it does not necessarily according to the regex. Curiously it seems smart-ish enough to deal with "hello\"world", which can only be seen as confusing and contradictory to the documentation of `*?`:

> The '*', '+', and '?' quantifiers are all greedy; they match as much text as possible. Sometimes this behaviour isn’t desired; if the RE <.*> is matched against '<a> b <c>', it will match the entire string, and not just '<a>'. Adding ? after the quantifier makes it perform the match in non-greedy or minimal fashion; as few characters as possible will be matched. Using the RE <.*?> will match only '<a>'.

https://docs.python.org/3/library/re.html

In this case, the `.` is allowed to match the `\`, and as such in `"hello\"world"`, to get "as few characters as possible", the option should be picked to get `"hello\"` (with `\` being matched by `.`) instead of `"hello\"world"`, with `\"` being matched by `\\"`.

Curiously enough, switching the `.` and `\\"` segments makes it so:

```
>>> re.search(r'"(?:\\"|.)*?"', '"hello\\"world"')
<re.Match object; span=(0, 14), match='"hello\\"world"'>
>>> re.search(r'"(?:.|\\")*?"', '"hello\\"world"')
<re.Match object; span=(0, 8), match='"hello\\"'>
```

Semantically, I think the regex should be:

```
r'"(?:\\.|[^\\"])*?"'
```

where the `*?` can even be replaced by `*`. This turns the regex to match a sequence of blocks that are either an escape-character followed by any other character (potentially a quote, but who cares), or a character that's neither a quote nor an escape-character.

Reply:
I'll have to some experimenting myself, but

> where the *? can even be replaced by *

Is likely wrong, because in `var = a "foo"; var b = "bar"` we wouldn't want to match `foo"; var b = "bar` in its entirety.

Reply:
Neither the `\\.` nor the `[^\\"]` parts will ever match the `"` in `foo"`. The first won't because it's not prefixed by `\`, the second won't because it refuses to match any `"`.

Reply:
After some experimentation, I removed one `"` character from you suggestion and ended up with `"(?:\\.|[^\\])*?"`:

```python
import re

regex = r'"(?:\\.|[^\\])*?"'

print(f"The regex `{regex}` does the following:")

should_match = [
    (r'"hello"', r'"hello"'),
    (r'"hello\"world"', r'"hello\"world"'),
    (r'"hello"world"', r'"hello"'),
    (r'"hello\\"world"', r'"hello\\"'),
]
should_not_match = [
    r'"hello\"',
    r'"hello\"world',
]

for should in should_match:
    if match := re.match(f"({regex})", should[0]):
        print(f"- from `{should[0]}` it catches `{should[1]}`, correct")
    else:
        print(f"- does not match `{should[0]}`, incorrect")

for should_not in should_not_match:
    if match := re.match(regex, should_not):
        print(f"- matches `{should_not}`, incorrect")
    else:
        print(f"- does not match `{should_not}`, correct")
```

Output:

The regex "(?:\\.|[^\\])*?" does the following:
- from "hello" it catches "hello", correct
- from "hello\"world" it catches "hello\"world", correct
- from "hello"world" it catches "hello", correct
- from "hello\\"world" it catches "hello\\", correct
- does not match "hello\", correct
- does not match "hello\"world, correct

Reply:
...aaaand now the unittests fail in a very, _very_ weird way:

```
  <!-- Quoted strings with escaped quotes -->
  <script>
      var foo = '\'{'
-     var bar = "\"{"
+             var bar = "\"{"
? ++++++++
      var baz = `\`{`
  </script>
```

Reply:
Further investigation revealed there is a difference between `re.match()` (which I used for experimenting) and `re.split()` (which DjHTML uses). I've changed the test code to this:

```python
from djhtml.modes import DjJS
token_re = DjJS(return_mode=True).token_re

print(f"The regex `{token_re}` does the following:")

should_match = [
    (r'"hello"', r'"hello"'),
    (r'"hello\"world"', r'"hello\"world"'),
    (r'"hello"world"', r'"hello"'),
    (r'"hello\\"world"', r'"hello\\"'),
    (r'"hello", "world"', r'"hello"'),
]
should_not_match = [
    r'"hello\"',
    r'"hello\"world',
]

for should in should_match:
    if match := token_re.match(should[0]):
        print(f"- from `{should[0]}` it catches `{should[1]}`, correct")
    else:
        print(f"- does not match `{should[0]}`, incorrect")

for should_not in should_not_match:
    if match := token_re.match(should_not):
        print(f"- matches `{should_not}`, incorrect")
    else:
        print(f"- does not match `{should_not}`, correct")

print("\nHowever, doing the same thing with re.split():")

for should in should_match:
    head, raw_token, tail = token_re.split(should[0], maxsplit=1)
    if raw_token == should[1]:
        print(f"- from `{should[0]}` it catches `{should[1]}`, correct")
    else:
        print(f"- from `{should[0]}` it catches `{raw_token}`, incorrect")

for should_not in should_not_match:
    try:
        head, raw_token, tail = token_re.split(should_not[0], maxsplit=1)
        print(f"- from `{should_not}` it catches `{raw_token}`, incorrect")
    except:
        print(f"- does not match `{should_not}`, correct")
```

And the output is:

The regex `re.compile('(\\n|{%[-\\+]?.*?[-\\+]?%}|{#.*?#}|{#|{{.*?}}|\\\\.|//.*|/\\*|[\\$\\w-]+:|"(?:\\\\.|[^\\\\])*"|\'(?:\\\\.|[^\\\\])*\'|`(?:\\\\.|[^\\\\])*`|`|[\\{\\[\\(\\)\\]\\}]|var |let |const |if(?= *\\()|else(?= )` does the following:
- from `"hello"` it catches `"hello"`, correct
- from `"hello\"world"` it catches `"hello\"world"`, correct
- from `"hello"world"` it catches `"hello"`, correct
- from `"hello\\"world"` it catches `"hello\\"`, correct
- from `"hello", "world"` it catches `"hello"`, correct
- does not match `"hello\"`, correct
- does not match `"hello\"world`, correct

However, doing the same thing with re.split():
- from `"hello"` it catches `"hello"`, correct
- from `"hello\"world"` it catches `"hello\"world"`, correct
- from `"hello"world"` it catches `"hello"world"`, incorrect
- from `"hello\\"world"` it catches `"hello\\"world"`, incorrect
- from `"hello", "world"` it catches `"hello", "world"`, incorrect
- does not match `"hello\"`, correct
- does not match `"hello\"world`, correct

It doesn't happen every day that I say this, but I think I found a bug in Python!

Reply:
Never mind, the test code itself contains a bug. Instead of

```python
print(f"- from `{should[0]}` it catches `{should[1]}`, correct")
```

It should have been

```python
print(f"- from `{should[0]}` it catches `{match[1]}`, correct")
```

And then you'll see the result of `split()` and `match()` are actually identical, and also both wrong. Changing the expression to the non-greedy variant `"(?:\\.|[^\\])*?"` solves it.

Reply:
Changing the expression to `"(?:\\.|[^\\"])*"` also solves it, and doesn't require non-greediness, so I've changed it to that. Case closed.